connmgr/conngating.go

```diff
@@ -0,0 +1,71 @@
+// connection. Hence, it is important to implement this interface keeping in mind the specific life-cycle state
+// at which you'd like to gate/block the connection.
+//
+// `InterceptDial` and `InterceptPeerDial` are called when we try an outbound dial.
+//
+// `InterceptAccept` is called after we've accepted an inbound connection from a socket but before we
+// begin upgrading it.
+//
+// `InterceptSecured` is called for inbound and outbound connections after we've negotiated
+// the security protocol to use for the connection.
+//
+// `InterceptUpgraded` is called for inbound and outbound connections after we've finished upgrading
+// a connection to have both security and stream multiplexing.
+//
+// This feature can be used to implement *strict* connection management
+// behaviours, such as hard limiting of connections once a max count has been
+// reached.
+//
+// If you'd like to send a disconnect control message to the remote peer for a gated inbound connection,
+// ONLY `InterceptUpgraded` should reject the connection (with an appropriate disconnect reason).
+// All other methods should allow the connection as we can ONLY open control streams
+// for upgraded connections.
+// Note: There's no point in sending disconnect control messages for outbound connections, so we might
+// as well close them as early in the cycle as possible.
+type ConnectionGater interface {
+	// InterceptDial tests whether we're permitted to dial the specified multiaddr.
+	// Insofar filter.Filters is concerned, this would map to its AddrBlock method,
+	// with the inverse condition.
+	// This is to be called by the network/swarm when dialling.
+	InterceptDial(ma.Multiaddr) (allow bool)
+
+	// InterceptPeerDial tests whether we're permitted to Dial the specified peer.
+	// This is to be called by the network/swarm when dialling.
+	InterceptPeerDial(p peer.ID) (allow bool)
+
+	// InterceptAccept tests whether an incipient inbound connection is allowed.
+	// network.ConnMultiaddrs is what we pass to the upgrader.
+	// This is intended to be called by the upgrader, or by the transport
+	// directly (e.g. QUIC, Bluetooth), straight after it's accepted a connection
+	// from its socket.
+	InterceptAccept(network.ConnMultiaddrs) (allow bool)
+
+	// InterceptSecured tests whether a given connection, now authenticated,
+	// is allowed.
+	// This is intended to be called by the upgrader, after it has negotiated crypto,
+	// and before it negotiates the muxer, or by the directly by the transport,
+	// at the exact same checkpoint.
+	InterceptSecured(network.Direction, peer.ID, network.ConnMultiaddrs) (allow bool)
+
+	// InterceptUpgraded tests whether a fully capable connection is allowed.
+	// At this point, we have a multiplexer, so the middleware can
+	// return a DisconnectReason and the swarm would use the control stream to convey
+	// it to the peer.
+	InterceptUpgraded(transport.CapableConn) (allow bool, reason control.DisconnectReason)
```

Comment:
Given that this is the "final check", why not do this in the swarm and use a normal connection? CapableConn has methods like `AcceptStream` that should only be called by the swarm, as far as I know.

Reply:
@Stebalien We do this in the Swarm ONLY:
https://github.com/libp2p/go-libp2p-swarm/blob/316a8557fa84f479d2a247ea4aaf1d7c5fc83812/swarm.go#L195

I'll update the docs to reflect this.

Reply:
Ah got it. In that case, I'd definitely use the normal `network.Conn` type.

Reply:
Agree. Changed here: https://github.com/libp2p/go-libp2p-core/pull/139/commits/32264f2111d56b982f47bb6623992e3e78710e89.